lib/explorer/polars_backend/lazy_data_frame.ex

```diff
@@ -0,0 +1,87 @@
+defmodule Explorer.PolarsBackend.LazyDataFrame do
+  @moduledoc false
+
+  alias Explorer.DataFrame
+  alias Explorer.PolarsBackend.Shared
+
+  @type t :: %__MODULE__{resource: binary(), reference: reference()}
+
+  defstruct resource: nil, reference: nil
+
+  use Explorer.Backend.DataFrame, backend: "LazyPolars"
+
+  # Conversion
+
+  @impl true
+  def lazy, do: __MODULE__
+
+  @impl true
+  def to_lazy(ldf), do: ldf
+
+  @impl true
+  def collect(ldf), do: Shared.apply_native(ldf, :lf_collect)
+
+  # Introspection
+
+  @impl true
+  def names(ldf), do: Shared.apply_native(ldf, :lf_names)
+
+  @impl true
+  def dtypes(ldf),
+    do:
+      ldf
+      |> Shared.apply_native(:lf_dtypes)
+      |> Enum.map(&Shared.normalise_dtype/1)
+
+  @impl true
+  def shape(ldf), do: {nil, n_columns(ldf)}
+
+  @impl true
+  def n_columns(ldf), do: ldf |> names() |> length()
+
+  # Single table verbs
+
+  @impl true
+  def head(ldf, rows), do: Shared.apply_native(ldf, :lf_head, [rows])
+
+  @impl true
+  def tail(ldf, rows), do: Shared.apply_native(ldf, :lf_tail, [rows])
+
+  @impl true
+  def select(ldf, columns, :keep) when is_list(columns),
+    do: Shared.apply_native(ldf, :lf_select, [columns])
+
+  def select(ldf, columns, :drop) when is_list(columns),
+    do: Shared.apply_native(ldf, :lf_drop, [columns])
+
+  @impl true
+  def pull(ldf, name), do: Shared.apply_native(ldf, :lf_pull, [name])
```

Comment:
I feel like we don't  want this? We always want to collect before we pull?

Reply:
It depends on how we define `pull`. I can be convinced either way. `pull` in `dplyr` always returns the full computed vector. We don't have to stick to `dplyr` but I can't see much value in users working with 'lazy' `Series` directly. Only in the context of DF operations.

Reply:
I agree but there's no way to get the realised series from a lazy series without some very major changes (keep the lazy df resource in the series struct?).

Reply:
You could introduce a LazySeries with a different backend implementation that keeps the dataframe but I think we should not allow the operation altogether (i.e. do not implement pull and let it raise instead). Especially because even if we can memoize/cache for LazyPolars, we won't be able to do it for ExplorerSQL? So I would postpone this!

Reply:
I would think we could memoize/cache for ExplorerSQL. dbplyr does it! But that's fine. It means either no Access or rewriting access and writing a custom inspect which mostly just copies the default implementation, but again not a huge deal. Also means that if you want to pull you need to select, then collect, then pull. Not the end of the world.

Reply:
> I would think we could memoize/cache for ExplorerSQL.

The problem is that there is no "easy" way to do this in Erlang/Elixir, I believe. We would need a process for coordination or some sort of special NIF. I will keep thinking about this problem though!